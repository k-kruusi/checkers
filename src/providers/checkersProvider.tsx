import { ReactNode, createContext, useMemo, useReducer } from 'react';
import { BoardState } from '../schema';
import { Action, reducer } from '../reducer';
import { Timer, cloneState } from '../utils';

export const CheckersContext = createContext<{
  state: BoardState;
  dispatch: React.Dispatch<Action>;
  blackTimer: Timer;
  redTimer: Timer;
} | null>(null);

export const CheckersProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, cloneState());

  const blackTimer = useMemo(() => new Timer(), []);
  const redTimer = useMemo(() => new Timer(), []);

  return (
    <CheckersContext.Provider value={{ state, dispatch, blackTimer, redTimer }}>
      {children}
    </CheckersContext.Provider>
  );
};
