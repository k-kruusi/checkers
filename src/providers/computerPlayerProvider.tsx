import { ReactNode, createContext, useCallback, useEffect, useState } from 'react';
import { useCheckers, useNextMove } from '../hooks';
import { getPieces, getRandomIndex, isTurn } from '../utils';
import { ComputerPlayerOption, MoveResult, Outcome, Piece, PlayerType } from '../schema';
import { ActionType } from '../reducer';

export const ComputerPlayerContext = createContext<{
  pickMove: (piece: Piece) => void;
} | null>(null);

export const ComputerPlayerProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const { state, dispatch } = useCheckers();
  const { board, turn, players, winner } = state;
  const { inspectAndReturn } = useNextMove();
  const [lastTurnRed, setLastTurnRed] = useState<number>(-1);
  const [lastTurnBlack, setLastTurnBlack] = useState<number>(-1);

  const pickMove = useCallback((piece: Piece) => {
    const { red, black } = getPieces(board);
    const movesWithoutJumps: ComputerPlayerOption[] = [];
    const movesWithJumps: ComputerPlayerOption[] = [];

    const piecesOfFocus = piece === Piece.Black ? black : red;

    if (piecesOfFocus.length === 0) {
      return null;
    }

    piecesOfFocus.forEach((tileData) => {
      let moves: Outcome[] = [];

      try {
        moves = inspectAndReturn(tileData);
        if (moves.length === 0) {
          return;
        }
      } catch (e) {
        console.error("Error calculating moves for computer opponent", e);
        return;
      }

      const numWithJumps = moves.filter((m) => m.didJump);
      if (numWithJumps.length > 0) {
        // always picking the last jump in the list
        // may not be the biggest jump but it should be one that cannot continue.
        // room to improve...
        movesWithJumps.push({ tileData, move: numWithJumps[numWithJumps.length - 1] });
      }
      else {
        // picking a random move from that tiles list.
        const index = getRandomIndex(moves);
        movesWithoutJumps.push({ tileData, move: moves[index] });
      }
    });

    if (movesWithJumps.length > 0) {
      const index = getRandomIndex(movesWithJumps);
      return movesWithJumps[index];
    }
    if (movesWithoutJumps.length > 0) {
      const index = getRandomIndex(movesWithoutJumps);
      return movesWithoutJumps[index];
    }

    return null;
  }, [inspectAndReturn, board]);


  // observes & triggers red if computer player
  useEffect(() => {
    if (lastTurnRed === turn.count || !isTurn(turn.phase, Piece.Red) || winner) {
      return;
    }
    const selection = pickMove(Piece.Red);
    if (!selection) {
      // game over there are no moves left for the computer.
      dispatch({ type: ActionType.FORFEIT, loser: Piece.Red });
      return;
    }

    const computerPlayer = players.find((p) => p.type === PlayerType.Computer);
    if (!computerPlayer) {
      return;
    }
    setLastTurnRed(turn.count);
    dispatch({
      type: ActionType.MOVE_PIECE,
      from: selection.tileData.coord,
      to: selection.move.coord,
      piece: selection.tileData.piece,
      potentials: [selection.move],
      result: selection.move.didJump ? MoveResult.Jump : MoveResult.Shift,
      timestamp: new Date().toISOString(),
    });
  }, [
    turn.phase, players, pickMove,
    dispatch, lastTurnRed, setLastTurnRed,
    turn.count, winner
  ]);

  // observes black
  useEffect(() => {
    if (lastTurnBlack === turn.count || !isTurn(turn.phase, Piece.Black) || winner) {
      return;
    }
    const selection = pickMove(Piece.Black);
    if (!selection) {
      dispatch({ type: ActionType.FORFEIT, loser: Piece.Black });
    }
    setLastTurnBlack(turn.count);
  }, [lastTurnBlack, turn, winner, pickMove, dispatch]);

  return (
    <ComputerPlayerContext.Provider value={{ pickMove }}>
      {children}
    </ComputerPlayerContext.Provider>
  );
};
