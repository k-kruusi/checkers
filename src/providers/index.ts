export * from './checkersProvider';
export * from './computerPlayerProvider';
export * from './nextMoveInspectorProvider';