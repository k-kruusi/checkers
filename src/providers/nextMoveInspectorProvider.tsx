import { createContext, useState } from 'react';
import { Outcome, Piece, TileData } from '../schema';
import { useCheckers } from '../hooks';
import { calculatePotentialMoves } from '../utils';

export const NextMoveInspectorContext = createContext<{
  inspect: (maybe: TileData) => void;
  inspectAndReturn: (maybe: TileData) => Outcome[];
  potentials: Outcome[];
  clear: () => void;
} | null>(null);

export const NextMoveInspectorProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const [focus, setFocus] = useState<TileData | null>(null);
  const [potentials, setPotentials] = useState<Outcome[]>([]);
  const { state } = useCheckers();

  const inspect = (fresh: TileData) => {
    if (focus &&
      focus.piece === fresh.piece &&
      focus.coord.x === fresh.coord.x &&
      focus.coord.y === fresh.coord.y) {
      // exit early already calculated for that tile.
      return;
    }
    // if you are hovering over an empty
    // flush the potentials
    if (fresh.piece === Piece.Empty) {
      setPotentials([]);
      setFocus(fresh);
      return;
    }

    try {
      const moves = calculatePotentialMoves(state.board, fresh.coord);
      setPotentials(moves);
      setFocus(fresh);
    } catch (e) {
      console.error('error calculating potential moves', e);
    }
  };

  const inspectAndReturn = (fresh: TileData): Outcome[] => {
    try {
      return calculatePotentialMoves(state.board, fresh.coord);
    } catch (e) {
      console.error('error inspecting possible moves');
      return [];
    }
  }

  const clear = () => setPotentials([]);

  return (
    <NextMoveInspectorContext.Provider value={{ inspect, inspectAndReturn, potentials, clear }}>
      {children}
    </NextMoveInspectorContext.Provider>
  );
};


