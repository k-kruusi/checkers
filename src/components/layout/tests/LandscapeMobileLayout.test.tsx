import { render } from '@testing-library/react';
import { CheckersProvider, NextMoveInspectorProvider } from '../../../providers';
import { LandscapeMobileLayout } from '../LandscapeMobileLayout';

describe('HorizontalLayout', () => {
  it('matches the snapshot', () => {
    const { asFragment } = render(
      <CheckersProvider>
        <NextMoveInspectorProvider>
          <LandscapeMobileLayout>
            <div>placeholder</div>
          </LandscapeMobileLayout>
        </NextMoveInspectorProvider>
      </CheckersProvider>);

    expect(asFragment()).toMatchSnapshot();
  });
});