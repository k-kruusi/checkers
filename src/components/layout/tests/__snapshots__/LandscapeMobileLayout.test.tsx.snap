// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HorizontalLayout matches the snapshot 1`] = `
<DocumentFragment>
  <div
    style="display: flex; flex-direction: row; justify-content: space-evenly; align-items: center;"
  >
    <div
      style="display: flex; flex-direction: column; justify-content: space-evenly; align-items: center; gap: 20px;"
    >
      <div
        style="text-align: center; color: rgb(255, 255, 240); font-size: 4em; font-family: Amatic SC, sans-serif;"
      >
        <h1>
          Birdseye
        </h1>
      </div>
      <button
        aria-label="reset"
        class="shared-button"
      >
        Reset
      </button>
      <div
        style="border: 4px solid #daa520; border-radius: 5px; min-width: 85px; user-select: none; font-family: Amatic SC, sans-serif;"
      >
        <div
          style="margin: 5px; padding: 5px 10px; background-color: rgb(255, 255, 240); font-weight: bold; text-align: left;"
        >
          <div>
            <h1
              style="font-size: 18px; text-align: center;"
            >
              Black
            </h1>
          </div>
          <div
            style="display: flex; flex-direction: row; align-items: center; gap: 5px; text-align: left; font-size: 18px;"
          >
            <svg>
              timer.svg
            </svg>
             
            <p>
              0:00
            </p>
          </div>
          <div
            style="display: flex; flex-direction: row; align-items: center; gap: 5px; text-align: left; font-size: 18px;"
          >
            <svg>
              pieces.svg
            </svg>
             
            <p>
              12
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      style="display: flex; flex-direction: column; justify-content: center; align-items: center; transform: translateY(10px);"
    >
      <div>
        placeholder
      </div>
    </div>
    <div
      style="display: flex; flex-direction: column; justify-content: space-evenly; align-items: center; gap: 20px;"
    >
      <div
        style="text-align: center; color: rgb(255, 255, 240); font-size: 4em; font-family: Amatic SC, sans-serif;"
      >
        <h1>
          Checkers
        </h1>
      </div>
      <button
        aria-label="ai-toggle"
        class="shared-button"
      >
        AI on
      </button>
      <div
        style="border: 4px solid #daa520; border-radius: 5px; min-width: 85px; user-select: none; font-family: Amatic SC, sans-serif;"
      >
        <div
          style="margin: 5px; padding: 5px 10px; background-color: rgb(70, 70, 70); font-weight: bold; text-align: left;"
        >
          <div>
            <h1
              style="font-size: 18px; text-align: center;"
            >
              Red
            </h1>
          </div>
          <div
            style="display: flex; flex-direction: row; align-items: center; gap: 5px; text-align: left; font-size: 18px;"
          >
            <svg>
              timer.svg
            </svg>
             
            <p>
              0:00
            </p>
          </div>
          <div
            style="display: flex; flex-direction: row; align-items: center; gap: 5px; text-align: left; font-size: 18px;"
          >
            <svg>
              pieces.svg
            </svg>
             
            <p>
              12
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
